notebooks/contrib-dev/Alcohol_Pharmacokinetics.ipynb

```diff
@@ -0,0 +1,1075 @@
+        "\n",
+        "    # Avoid the first time\n",
+        "    if below_limit_indices.size > 0:\n",
+        "        # Add first_above_index to it to get the index in the original array\n",
+        "        first_below_index = below_limit_indices[0] + first_above_index\n",
+        "\n",
+        "        # Find the time\n",
+        "        time_falls_below = C[4][first_below_index]\n",
+        "        print(f\"Time to sober up to below the legal limit: {time_falls_below:.2f} hours\")\n",
+        "    else:\n",
+        "        print(\"BAC does not fall below the legal limit within the time frame.\")\n",
+        "else:\n",
+        "    print(\"BAC never exceeds the legal limit within the time frame.\")\n",
+        "\n",
+        "\n",
+        "### END SOLUTION"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "C3HXfCDa2-M0"
+      },
+      "source": [
+        "Using the plotted kinetics **Discuss in 3-6 sentences** how long it takes for alcohol to enter the blood stream vs how long it takes for the alcohol to break down. How would BAC change over time if the person weighed more/less? If they drink more/less?\n"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "VK-3NOub6R64"
+      },
+      "source": [
+        "**Discussion:**"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## 1D. Example of Forward Euler Method for ODEs\n",
+        "\n",
+        "The Forward Euler method is a straightforward numerical approach for solving ODEs.\n",
+        "\n",
+        "Given the Pharmacokineticsm, simplify it and and consider a first-order linear ODE of the form:\n",
+        "\n",
+        "$$\n",
+        "\\frac{dC}{dt} = -kC\n",
+        "$$\n",
+        "\n",
+        "where ( C ) represents the concentration of a substance (e.g., alcohol in the blood), and ( k ) is the rate constant for elimination.\n",
+        "\n",
+        "The Forward Euler method approximates the derivative at a point by considering the slope of the line connecting the current point and the next point, which can be expressed as:\n",
+        "\n",
+        "$$\n",
+        "\\frac{dC}{dt} \\approx \\frac{C(t+\\Delta t) - C(t)}{\\Delta t}\n",
+        "$$\n",
+        "\n",
+        "Using this approximation, we can update the concentration ( C ) at each time step ( ${\\Delta t}$ ) as follows:\n",
+        "\n",
+        "$$\n",
+        "C(t+\\Delta t) = C(t) - kC(t)\\Delta t\n",
+        "$$\n",
+        "\n",
+        "This method iteratively updates the value of ( C ) over time, allowing us to model the change in concentration due to the process described by the ODE."
+      ],
+      "metadata": {
+        "id": "bW6nx9fvchxy"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "def forward_euler(k, C0, tmax, dt):\n",
```

Comment:
This function combines the logic for Euler's method with the specifics of the model. This is NOT best practice. Instead, we want to separate the logic for the general numerical method from the model details. Please see examples on the class website. Alternately, please remove Euler's method to instead focus on comparing the models.